jerry-core/api/jerry.c

```diff
@@ -2463,6 +2463,92 @@ jerry_set_object_native_handle (const jerry_value_t obj_val, /**< object to set
   }
 } /* jerry_set_object_native_handle */
 
+/**
+ * Traverse objects.
+ *
+ * @return true - traversal was interrupted by the callback.
+ *         false - otherwise - traversal visited all objects.
+ */
+bool jerry_objects_foreach (jerry_objects_foreach_t foreach_p,
+                            void *user_data_p)
+{
+  jerry_assert_api_available ();
+
+  ecma_object_t *iter_p;
+
+  JERRY_ASSERT (foreach_p != NULL);
+
+  for (iter_p = JERRY_CONTEXT (ecma_gc_objects_p);
+       iter_p != NULL;
+       iter_p = ECMA_GET_POINTER (ecma_object_t, iter_p->gc_next_cp))
+  {
+    if (!ecma_is_lexical_environment (iter_p)
+        && !foreach_p (ecma_make_object_value (iter_p), user_data_p))
+    {
+      return true;
+    }
+  }
+
+  return false;
+} /* jerry_objects_foreach */
+
+typedef struct
+{
+  jerry_objects_foreach_by_native_info_t foreach_p;
+  const jerry_object_native_info_t *native_info_p;
+  void *user_data_p;
+} jerry_by_native_iteration_info_t;
+
+static bool
+jerry_objects_foreach_by_native_info_callback (const jerry_value_t value,
+                                               void *user_data_p)
+{
+  jerry_by_native_iteration_info_t *iteration_info_p = (jerry_by_native_iteration_info_t *) user_data_p;
+  ecma_native_pointer_t *native_pointer_p = ecma_get_native_pointer_value (ecma_get_object_from_value (value),
+                                                                           LIT_INTERNAL_MAGIC_STRING_NATIVE_POINTER);
+
+  /* Keep iterating if the object has no native info. */
+  if (native_pointer_p == NULL)
+  {
+    return true;
+  }
+
+  /* Keep iterating if the object's native info is not of the type we're looking for. */
+  if (((const jerry_object_native_info_t *) native_pointer_p->u.info_p) != iteration_info_p->native_info_p)
+  {
+    return true;
+  }
+
+  /* We found an object with the desired type of native info. Let the callback decide whether to keep iterating. */
+  return iteration_info_p->foreach_p (value, native_pointer_p->data_p, iteration_info_p->user_data_p);
+} /* jerry_objects_foreach_by_native_info_callback */
+
+/**
+ * Traverse objects having a given native type info.
+ *
+ * @return true - traversal was interrupted by the callback.
+ *         false - otherwise - traversal visited all objects.
+ */
+bool
+jerry_objects_foreach_by_native_info (const jerry_object_native_info_t *native_info_p,
+                                      jerry_objects_foreach_by_native_info_t foreach_p,
+                                      void *user_data_p)
+{
+  jerry_assert_api_available ();
+
+  JERRY_ASSERT (native_info_p != NULL);
+  JERRY_ASSERT (foreach_p != NULL);
+
+  jerry_by_native_iteration_info_t iteration_info =
+  {
+    .foreach_p = foreach_p,
+    .native_info_p = native_info_p,
+    .user_data_p = user_data_p
+  };
+
+  return jerry_objects_foreach (jerry_objects_foreach_by_native_info_callback, &iteration_info);
+} /* jerry_objects_foreach_by_native_info */
```

Comment:
Maybe my impression but using the other function to implement this actually increase complexity rather than decrease it. I would just duplicate the code.